orttraining/orttraining/core/optimizer/lstm_replacement.cc

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/optimizer/lstm_replacement.h"
+
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status LSTMReplacement::Apply(Graph& graph, Node& lstm_node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  const auto& lstm_inputs = lstm_node.MutableInputDefs();
+  auto& lstm_outputs = lstm_node.MutableOutputDefs();
+
+  ONNX_NAMESPACE::TypeProto lstm_input_type;
+  lstm_input_type.mutable_tensor_type()->set_elem_type(lstm_inputs.front()->TypeAsProto()->tensor_type().elem_type());
+
+  if (lstm_outputs.empty() || !lstm_outputs[0]->Exists()) {
+    // Add all LSTM cell optional outputs for computation even if not required to ensure all outputs are computed
+    // so that gradient computation can pick them up.
+    NodeArg& all_hidden_states = graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("all_hidden_states"),
+                                                          &lstm_input_type);
+    if (!lstm_outputs.empty()) {
+      lstm_outputs[0] = &all_hidden_states;
+    } else {
+      lstm_outputs.push_back(&all_hidden_states);
+    }
+  }
+
+  if (lstm_outputs.size() == 1U) {
+    NodeArg& placeholder = graph.GetOrCreateNodeArg("", nullptr);
+    lstm_outputs.push_back(&placeholder);
+  }
+
+  if (lstm_outputs.size() == 2U) {
+    NodeArg& placeholder = graph.GetOrCreateNodeArg("", nullptr);
+    lstm_outputs.push_back(&placeholder);
+  }
+
+  NodeArg& all_cell_states = graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("all_cell_states"),
```

Comment:
>all_cell_states [](http://example.com/codeflow?start=81&length=15)

once we create such an output in forward, who will use it? then how the all_cell_states's gradient get generated then?

Reply:
`all_cell_states` gets only consumed by the LSTMGrad op. Gradients are not generated for this tensor since no other forward op consumes this tensor.
It is an intermediate stashed tensor that the gradient kernel needs to compute gradients.